### PR TITLE
[data grid] Fixed issue where pressing Delete key resets various cell values to empty string.

### LINKED ISSUE #12125: [data grid] Pressing `Delete` key in a Boolean type cell, the value incorrectly resets to empty string
### Steps to reproduce

Link to live example: https://codesandbox.io/p/github/mororyou/mui-datagrid-test/main

1. To open Developer Tools

2. Hover over the cells in the isAdmin column.
   
3. To delete, use the Delete key or the Backspace key.

4. Check the Console in Developer Tools.
   
5. You can confirm in the console that the value of valueSetter's params is an empty string.

Also, I'll attach a screenshot for reference.
https://gyazo.com/787beae18b71f4366be716d654a6cf50

### Current behavior

I was expecting to receive a Boolean, but an empty string is being sent.

### Expected behavior

Can we change the Boolen value so that it can be retrieved?

### Context

_No response_

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Search keywords**: DataGrid, valueSetter
**Order ID**: 67619

### PR DESCRIPTION
**Pull request by sooster910**
Fix #12125  [michelengelen](https://github.com/michelengelen)

This PR refines the value reset logic to align with column data types on Backspace/Delete key press.

UnitTesting
Implemented tests to check cell value reset upon Backspace and Delete key presses.

The video shows the correct default boolean value when the delete/backspace key is pressed.

mui-x

### CODE REVIEW

**Review by michelengelen — APPROVED** [review]
LGTM ... thanks for picking it up!

Let's wait for the review from @romgrk though!

**Comment by sooster910** [comment]
> LGTM ... thanks for picking it up!
> 
> Let's wait for the review from @romgrk though!

Thanks a lot!!

**Review by romgrk — APPROVED** [review]

**Pull request merged**
